Covid19Radar/Covid19Radar/Repository/ExposureConfigurationRepository.cs

```diff
@@ -0,0 +1,340 @@
+        private readonly ILocalPathService _localPathService;
+        private readonly IPreferencesService _preferencesService;
+        private readonly IServerConfigurationRepository _serverConfigurationRepository;
+        private readonly IDateTimeUtility _dateTimeUtility;
+        private readonly ILoggerService _loggerService;
+
+        private readonly string _configDir;
+
+        private string _currentExposureConfigurationPath;
+
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+
+        public ExposureConfigurationRepository(
+            IHttpClientService httpClientService,
+            ILocalPathService localPathService,
+            IPreferencesService preferencesService,
+            IServerConfigurationRepository serverConfigurationRepository,
+            IDateTimeUtility dateTimeUtility,
+            ILoggerService loggerService
+            )
+        {
+            _client = httpClientService.Create();
+            _localPathService = localPathService;
+            _preferencesService = preferencesService;
+            _serverConfigurationRepository = serverConfigurationRepository;
+            _dateTimeUtility = dateTimeUtility;
+            _loggerService = loggerService;
+
+            _configDir = PrepareConfigDir();
+            _currentExposureConfigurationPath = localPathService.CurrentExposureConfigurationPath;
+        }
+
+        private string PrepareConfigDir()
+        {
+            var configDir = _localPathService.ExposureConfigurationDirPath;
+            if (!Directory.Exists(configDir))
+            {
+                Directory.CreateDirectory(configDir);
+            }
+            return configDir;
+        }
+
+        public async Task<ExposureConfiguration> GetExposureConfigurationAsync()
+        {
+            await _semaphore.WaitAsync();
+
+            try
+            {
+                return await GetExposureConfigurationInternalAsync();
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
+        }
+
+        private async Task<ExposureConfiguration> GetExposureConfigurationInternalAsync()
+        {
+            _loggerService.StartMethod();
+
+            ExposureConfiguration currentExposureConfiguration = null;
+
+            if (File.Exists(_currentExposureConfigurationPath))
+            {
+                _loggerService.Debug("ExposureConfiguration file is found.");
+
+                try
+                {
+                    string exposureConfigurationAsJson = await LoadAsync(_currentExposureConfigurationPath);
+
+                    currentExposureConfiguration = JsonConvert.DeserializeObject<ExposureConfiguration>(exposureConfigurationAsJson);
+
+                    if (!IsDownloadedExposureConfigurationOutdated(AppConstants.ExposureConfigurationFileDownloadCacheRetentionDays))
+                    {
+                        return currentExposureConfiguration;
```

Comment:
`_loggerService.EndMethod();` を呼び出してからリターンすべきです。

```suggestion
                        _loggerService.EndMethod();
                        return currentExposureConfiguration;
```

Reply:
6a3cbf0 で対応しました！